demos/isoline_simplification/isoline_simplification_demo.cpp

```diff
@@ -0,0 +1,682 @@
+	dyken_r->setValue(1);
+	dyken_r->setMinimum(0);
+	dyken_r->setMaximum(10);
+	auto* dyken_r_label = new QLabel("Dyken's parameter R");
+	dyken_r_label->setBuddy(dyken_r);
+	vLayout->addWidget(dyken_r_label);
+	vLayout->addWidget(dyken_r);
+
+	auto* step_button = new QPushButton("Step");
+	vLayout->addWidget(step_button);
+
+	auto* outputDirectorySelector = new QPushButton("Select output directory");
+	vLayout->addWidget(outputDirectorySelector);
+
+	auto* save_button = new QPushButton("Save");
+	vLayout->addWidget(save_button);
+
+	auto* debug_text = new QLabel("");
+	vLayout->addWidget(debug_text);
+
+	auto* measure_text = new QLabel("");
+	vLayout->addWidget(measure_text);
+
+	m_renderer = new GeometryWidget();
+	m_renderer->setDrawAxes(showGrid->checkState());
+	setCentralWidget(m_renderer);
+
+	m_renderer->setMinZoom(0.01);
+	m_renderer->setMaxZoom(1000.0);
+
+	m_save = [this, debugInfo, showVertices, fileSelector, collapse_selector, measure_text, applySmoothing]() {
+		if (!m_output_dir.has_value() || m_output_dir == "") {
+			QMessageBox msgBox;
+			msgBox.setText("Please select an output directory.");
+			msgBox.exec();
+			return;
+		}
+
+		auto& isolines = m_isoline_simplifier->m_simplified_isolines;
+	  	IpeRenderer ipe_renderer;
+	  	auto isolines_p = std::make_shared<IsolinePainting>(isolines, showVertices->isChecked(), false, true,
+		                                                    applySmoothing->isChecked());
+
+	  	if (debugInfo->isChecked()) {
+			auto separator_p = std::make_shared<MedialAxisSeparatorPainting>(
+			    m_isoline_simplifier->m_separator, m_isoline_simplifier->m_delaunay);
+			auto slope_ladder_p =
+			    std::make_shared<SlopeLadderPainting>(m_isoline_simplifier->m_slope_ladders);
+			auto matching_p =
+			    std::make_shared<CompleteMatchingPainting>(m_isoline_simplifier->m_matching);
+
+			auto vdp = std::make_shared<VoronoiExceptMedialPainting>(*m_isoline_simplifier);
+			auto map = std::make_shared<MedialAxisExceptSeparatorPainting>(*m_isoline_simplifier);
+
+			ipe_renderer.addPainting(vdp, "Voronoi");
+			ipe_renderer.addPainting(map, "Medial_axis");
+			if (!m_isoline_simplifier->m_started) {
+				ipe_renderer.addPainting(separator_p, "Separator");
+			}
+			ipe_renderer.addPainting(slope_ladder_p, "Slope_ladders");
+			ipe_renderer.addPainting(matching_p, "Matching");
+		}
+	    ipe_renderer.addPainting(isolines_p, "Simplified_isolines");
+
+		if (m_debug_ladder.has_value()) {
+			auto ladder = *m_debug_ladder;
+			auto debug_ladder_p = std::make_shared<DebugLadderPainting>(*m_isoline_simplifier, *ladder);
+			ipe_renderer.addPainting(debug_ladder_p, "Debug_ladder_painting");
+		}
+
+		std::string file_name(fileSelector->currentText().toStdString() + "_" +
+		                      std::to_string(m_isoline_simplifier->m_current_complexity) + "_" +
+		                      collapse_selector->currentText().toStdString());
+	   	ipe_renderer.save(*m_output_dir + "/" + file_name + ".ipe");
```

Comment:
It would be nice to use std::filesystem for path construction operations like these.

Reply:
I cleaned it up a bit now.